### PR TITLE
fix #13, Add Abandoned type to TransitionState enum

### LINKED ISSUE #13: Add Abandoned type to TransitionState enum
Raft is an asynchronous protocol, so one of the challenges when using it is getting feedback on whether a particular transition has or hasn't been applied. To solve this problem, Little Raft calls the user-defined `register_transition_state` hook every time any transition changes its state. This way the library user can keep track of transitions as they move through the `Queued -> Committed -> Applied` pipeline.

However, a transition could be ignored by the replica it has been submitted to. The most likely reason for that is because the replica is not the cluster leader. Another possible reason is that the replica used to be the leader, then got disconnected from the rest of the cluster, kept accepting transitions for processing for a while and then had to drop them when connecting back to the cluster that in the meanwhile elected another, newer leader (the stale leader dropping uncommitted transitions is the desired behavior in this case).

To let the user know when a transition got dropped, we should add `Abandoned` state to the `TransitionState` enum. We could also have that `Abandoned` state wrap around another type so that we could signal the particular reason why a transition has been abandoned -- `NotLeader` vs `UncommittedUnsynced`. Naming could be improved.

### PR DESCRIPTION
**Pull request by suzaku**

### CODE REVIEW

File: little_raft/src/state_machine.rs [hunk 1]
```diff
@@ -16,6 +16,15 @@ pub enum TransitionState {
...
     /// been applied to the local state machine.
     Applied,
+
+    /// Abandoned transitions have been ignored by the replica.
+    Abandoned(TransitionAbandonedReason),
+}
+
+#[derive(Clone, Debug, PartialEq)]
+pub enum TransitionAbandonedReason {
+    // NotLeader transitions have been abandoned because the replica is not the cluster leadedr.
```
Comment (penberg) [inline little_raft/src/state_machine.rs:26]: s/leadedr/leader/

File: little_raft/src/replica.rs [hunk 2]
```diff
@@ -377,6 +377,11 @@ where
                 let mut state_machine = self.state_machine.lock().unwrap();
                 state_machine
                     .register_transition_state(transition.get_id(), TransitionState::Queued);
+            } else {
```
Comment (andreev-io) [inline little_raft/src/replica.rs:380]: Looks good to me. Little Raft is single-threaded, so we don't have to worry about state changing under our feet, so let's move the Leader check on line 369 here out of the `for` loop for efficiency just like you did in the previous draft PR.

And then let's reuse `let mut state_machine` declaration by declaring the variable outside the `if` statement.

File: little_raft/src/state_machine.rs (hunk 1 above)
Comment (andreev-io) [reply little_raft/src/state_machine.rs:26]: Let's also keep comment width at 80 characters please.

**Review by andreev-io — CHANGES_REQUESTED** [review]

**Comment by suzaku** [comment]
@andreev-io Updated.

**Comment by penberg** [comment]
Looks good. Is this in draft mode on purpose, @suzaku, or can this be merged?

**Comment by suzaku** [comment]
> Looks good. Is this in draft mode on purpose, @suzaku, or can this be merged?

I would like to add a test case for this, but maybe we can do that in another PR if you don't mind.

**Review by penberg — APPROVED** [review]

**Review by andreev-io — APPROVED** [review]

**Pull request merged**